### ISSUE TITLE
Filtering a table using `==`  doesn't work when a ForeignKey is involved (and can silently return incorrect results)

### ISSUE THREAD

**Issue by dhirschfeld** [blaze/blaze#1552]
As shown below when using the `==` operator to filter rows only a single match is returned. In comparison, the `isin` operator correctly returns the full set of matching rows:
```python
In [16]: import blaze as bz

In [17]: bz.__version__
Out[17]: '0.10.1'

In [19]: T[T.cce_contractid.isin({12199})].nrows
Out[19]: 35088

In [20]: T[T.cce_contractid == 12199].nrows
Traceback (most recent call last):

  File "<ipython-input-20-22f63c35d212>", line 1, in <module>
    T[T.cce_contractid == 12199].nrows

  File "C:\Python\lib\site-packages\blaze\expr\expressions.py", line 222, in __getattr__
    result = object.__getattribute__(self, key)

AttributeError: 'Slice' object has no attribute 'nrows'


In [21]: odo(T[T.cce_contractid == 12199], pd.DataFrame).shape
C:\Python\lib\site-packages\blaze\compute\sql.py:1414: UserWarning: The order of the result set from a Slice expression computed against the SQL backend is not deterministic.
  warnings.warn('The order of the result set from a Slice expression '
Out[21]: (1, 20)
```

**Comment by dhirschfeld** [comment]
It appears that this only happens when the column used as a filter has a foreign key constraint and is thus defined as a map in datashape
**Repro:**
```python
from itertools import chain

import blaze as bz
import numpy as np
from numpy.random import choice, randint
from odo import odo
import pandas as pd
import sqlalchemy as sa

engine = sa.create_engine('sqlite://')
metadata = sa.MetaData(bind=engine)
T1 = sa.Table('demo', metadata,
   sa.Column('timestamp', sa.String, primary_key=True),
   sa.Column('label_id', sa.Integer, sa.ForeignKey("labels.label_id"), nullable=False),
)
T2 = sa.Table('labels', metadata,
   sa.Column(
       'label_id',
       sa.Integer,
       primary_key=True,
       nullable=False,
   ),
   sa.Column('label', sa.String),
)

metadata.create_all()


labels = pd.DataFrame(
    list(zip(range(4), ['A','B','C','D'])),
    columns=['label_id', 'label']
)
odo(labels, T2)


dates = pd.date_range('01-Feb-2016', '01-Mar-2016', freq='D')
df = pd.DataFrame({'timestamp': list(map(str, dates))})
df['label_id'] = randint(0, labels.shape[0], dates.size)
odo(df, T1)
data = bz.Data(T1)
```
```python
In [1]: data.schema
Out[1]: dshape("{timestamp: string, label_id: map[int32, {label_id: int32, label: ?string}]}")

In [2]: df[df.label_id == 2]
Out[2]: 
              timestamp  label_id
2   2016-02-03 00:00:00         2
15  2016-02-16 00:00:00         2
17  2016-02-18 00:00:00         2
18  2016-02-19 00:00:00         2
20  2016-02-21 00:00:00         2

In [3]: odo(data[data.label_id.isin({2})], pd.DataFrame)  # works correctly
Out[3]: 
             timestamp  label_id
0  2016-02-03 00:00:00         2
1  2016-02-16 00:00:00         2
2  2016-02-18 00:00:00         2
3  2016-02-19 00:00:00         2
4  2016-02-21 00:00:00         2

In [4]: odo(data[data.label_id == 2], pd.DataFrame)  # BUG!
Out[4]: 
             timestamp  label_id
0  2016-02-01 00:00:00         0
```

**Comment by dhirschfeld** [comment]
**New Bug:**
It's handy to be able to filter a table base on the label name rather than the id but when doing so using blaze you silently get the wrong results. Using the same repro above:
```python
In [12]: labels
Out[12]: 
   label_id label
0         0     A
1         1     B
2         2     C
3         3     D

In [13]: # this correctly filters on label 'C' when using the `label_id`
    ...: # Note: we can't use a simple `==` because of the previous bug
    ...: odo(data[data.label_id.isin({2})], pd.DataFrame)
Out[13]: 
             timestamp  label_id
0  2016-02-03 00:00:00         2
1  2016-02-06 00:00:00         2
2  2016-02-08 00:00:00         2
3  2016-02-09 00:00:00         2
4  2016-02-12 00:00:00         2
5  2016-02-16 00:00:00         2
6  2016-02-22 00:00:00         2
7  2016-02-27 00:00:00         2
8  2016-02-28 00:00:00         2

In [14]: # if we try to filter on the label itself we get nonsense
    ...: odo(data[data.label_id.label == 'C'], pd.DataFrame)  # BUG!!!
Out[14]: 
              timestamp  label_id
0   2016-02-01 00:00:00         3
1   2016-02-02 00:00:00         0
2   2016-02-03 00:00:00         2
3   2016-02-04 00:00:00         1
4   2016-02-05 00:00:00         1
5   2016-02-06 00:00:00         2
6   2016-02-07 00:00:00         1
7   2016-02-08 00:00:00         2
8   2016-02-09 00:00:00         2
9   2016-02-10 00:00:00         1
10  2016-02-11 00:00:00         0
11  2016-02-12 00:00:00         2
12  2016-02-13 00:00:00         3
13  2016-02-14 00:00:00         3
14  2016-02-15 00:00:00         0
15  2016-02-16 00:00:00         2
16  2016-02-17 00:00:00         3
17  2016-02-18 00:00:00         1
18  2016-02-19 00:00:00         0
19  2016-02-20 00:00:00         3
20  2016-02-21 00:00:00         0
21  2016-02-22 00:00:00         2
22  2016-02-23 00:00:00         0
23  2016-02-24 00:00:00         0
24  2016-02-25 00:00:00         0
25  2016-02-26 00:00:00         3
26  2016-02-27 00:00:00         2
27  2016-02-28 00:00:00         2
28  2016-02-29 00:00:00         1
29  2016-03-01 00:00:00         0

In [15]: # it fails with a NotImplementedError when using `isin`
    ...: odo(data[data.label_id.label.isin({'C'})], pd.DataFrame)  # BUG!!!
Traceback (most recent call last):

  File "<ipython-input-15-053c4aa972d9>", line 2, in <module>
    odo(data[data.label_id.label.isin({'C'})], pd.DataFrame)  # BUG!!!

  File "c:\dev\packages\odo\odo\odo.py", line 91, in odo
    return into(target, source, **kwargs)

  File "C:\Python\lib\site-packages\multipledispatch\dispatcher.py", line 164, in __call__
    return func(*args, **kwargs)

  File "C:\Python\lib\site-packages\blaze\interactive.py", line 404, in into
    result = compute(b, return_type='native', **kwargs)

  File "C:\Python\lib\site-packages\multipledispatch\dispatcher.py", line 164, in __call__
    return func(*args, **kwargs)

  File "C:\Python\lib\site-packages\blaze\interactive.py", line 195, in compute
    return compute(expr, resources, **kwargs)

  File "C:\Python\lib\site-packages\multipledispatch\dispatcher.py", line 164, in __call__
    return func(*args, **kwargs)

  File "C:\Python\lib\site-packages\blaze\compute\core.py", line 409, in compute
    result = top_then_bottom_then_top_again_etc(expr3, d4, **kwargs)

  File "C:\Python\lib\site-packages\blaze\compute\core.py", line 189, in top_then_bottom_then_top_again_etc
    return top_then_bottom_then_top_again_etc(expr3, scope4, **kwargs)

  File "C:\Python\lib\site-packages\blaze\compute\core.py", line 189, in top_then_bottom_then_top_again_etc
    return top_then_bottom_then_top_again_etc(expr3, scope4, **kwargs)

  File "C:\Python\lib\site-packages\blaze\compute\core.py", line 187, in top_then_bottom_then_top_again_etc
    "data: %s" % (type(expr3), expr3, scope4))

NotImplementedError: Don't know how to compute:
type(expr): <class 'blaze.expr.expressions.Selection'>
expr: _1[label.isin(frozenset({'C'}))]
data: {<`label` symbol; dshape='var * ?string'>: <sqlalchemy.sql.selectable.Select at 0xf66c9b0; Select object>, <`_1` symbol; dshape='var * {  timestamp: string,  label_id: map[int32, ...'>: Table('demo', MetaData(bind=Engine(sqlite://)), Column('timestamp', String(), table=<demo>, primary_key=True, nullable=False), Column('label_id', Integer(), ForeignKey('labels.label_id'), table=<demo>, nullable=False), schema=None)}
```

**Comment by dhirschfeld** [comment]
For reference the ML thread is at https://groups.google.com/a/continuum.io/forum/#!topic/blaze-dev/GGbog_WB4Do

**Comment by dhirschfeld** [comment]
After updating to master and the filtering with `isin` does seem to work. `==` still returns nonsense, but that is likely to the initial bug reported above
```python
In [5]: import blaze as bz

In [6]: bz.__version__
Out[6]: '0.11.1rc2'

In [7]: odo(data[data.label_id.label.isin({'C'})], pd.DataFrame)  # Now works on master! :)
Out[7]:
             timestamp  label_id
0  2016-02-05 00:00:00         2
1  2016-02-08 00:00:00         2
2  2016-02-16 00:00:00         2
3  2016-02-17 00:00:00         2
4  2016-02-20 00:00:00         2
5  2016-02-22 00:00:00         2
6  2016-02-25 00:00:00         2
7  2016-03-01 00:00:00         2

In [8]: odo(data[data.label_id.label == 'C'], pd.DataFrame)  # still broken
Out[8]:
              timestamp  label_id
0   2016-02-01 00:00:00         1
1   2016-02-02 00:00:00         3
2   2016-02-03 00:00:00         1
3   2016-02-04 00:00:00         3
4   2016-02-05 00:00:00         2
5   2016-02-06 00:00:00         3
6   2016-02-07 00:00:00         0
7   2016-02-08 00:00:00         2
8   2016-02-09 00:00:00         3
9   2016-02-10 00:00:00         0
..                  ...       ...
20  2016-02-21 00:00:00         1
21  2016-02-22 00:00:00         2
22  2016-02-23 00:00:00         1
23  2016-02-24 00:00:00         1
24  2016-02-25 00:00:00         2
25  2016-02-26 00:00:00         3
26  2016-02-27 00:00:00         3
27  2016-02-28 00:00:00         1
28  2016-02-29 00:00:00         3
29  2016-03-01 00:00:00         2

[30 rows x 2 columns]
```

**Comment by skrah** [comment]
@dhirschfeld  Thanks for the report and the detailed repro!   PR #1561 should fix the equality issue -- the ```_eq``` method simply wasn't discovered when the ```datashape``` was a ```Map```.

**Comment by kwmsmith** [comment]
@dhirschfeld now that #1561 has been merged into master, can you test again and let us know if your issue is resolved?

**Comment by dhirschfeld** [comment]
@skrah, not a problem - I'm highly motivated to make it as easy as possible for you to fix my bugs! :stuck_out_tongue_winking_eye: 

Thanks for the quick fix - i can confirm it works for me on master so am happy for this to be closed as fixed.

**Comment by kwmsmith** [comment]
Fixed by #1561.
